### PR TITLE
Better handle invalid parameters to signrawtransaction

### PR DESCRIPTION
**Pull request by TheBlueMatt**
Based on #9634.

I remembered why I wrote that to begin with - I was trying to merge signatures from untrusted parties and realized they could cause a DoS by crashing my bitcoind.

This silently skips trying to merge signatures from inputs which
do not exist from transactions provided to signrawtransaction,
instead of hitting an assert.

### CODE REVIEW

**Comment by TheBlueMatt** [comment]
Add test cases here for both this PR and #9634. The test fail if either is reverted.

File: qa/rpc-tests/signrawtransactions.py [hunk 1]
```diff
@@ -78,6 +78,20 @@ def script_verification_error_test(self):
...
         rawTx = self.nodes[0].createrawtransaction(inputs, outputs)
+
+        # Make sure decoderawtransaction is at least marginally sane
+        decodedRawTx = self.nodes[0].decoderawtransaction(rawTx)
+        for i, inp in enumerate(inputs):
+            assert_equal(decodedRawTx["vin"][i]["txid"], inp["txid"])
+            assert_equal(decodedRawTx["vin"][i]["vout"], inp["vout"])
+
+        # Make sure decoderawtransaction throws if there is extra data
+        try:
```
Comment (MarcoFalke) [inline qa/rpc-tests/signrawtransactions.py]: Nit: The following is usually referred to as `assert_raises(JSONRPCException, self.nodes...

File: qa/rpc-tests/signrawtransactions.py [hunk 2]
```diff
@@ -48,6 +48,22 @@ def successful_signing_test(self):
...
         assert 'errors' not in rawTxSigned
 
+        # Check that signrawtransaction doesn't blow up on garbate merge attempts
+        dummyTxInconsistent = self.nodes[0].createrawtransaction([inputs[0]], outputs)
+        rawTxUnsigned = self.nodes[0].signrawtransaction(rawTx + dummyTxInconsistent, inputs)
+
+        assert 'complete' in rawTxUnsigned
+        assert_equal(rawTxUnsigned['complete'], False)
+
+        # Check that signrawtransaction properly merges unsigned and signed txn, even with garbate in the middle
```
Comment (jtimon) [inline qa/rpc-tests/signrawtransactions.py]: s/garbate/garbage/ ?

**Comment by TheBlueMatt** [comment]
Fixed review nits.

**Comment by TheBlueMatt** [comment]
Fixed another "garbage" misspelling....

**Comment by laanwj** [comment]
utACK 6dbfe08

**Comment by jnewbery** [comment]
Apologies - post merge question/comment. The new code in DecodeHexTx adds this check for witness transactions not having junk bytes:

```
        if (!ssData.empty())
            return false;
```

12 lines above is a check for non-witness transactions not having junk bytes:

```
            if (ssData.eof()) {
                return true;
```

I'm pretty sure that `ssData.eof()` and `ssData.empty()` will both be false if there are junk bytes remaining in the CDataStream.

Any reason why you've used a different function call? It's a bit confusing for someone reading the code for the first time.

**Pull request merged**